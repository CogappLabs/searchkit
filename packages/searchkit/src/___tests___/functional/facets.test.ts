import Client, { AlgoliaMultipleQueriesQuery } from '../..'
import nock from 'nock'
import { DisjunctiveExampleRequest } from '../mocks/AlgoliaRequests'
import { HitsResponseWithFacetFilter } from '../mocks/ElasticsearchResponses'

describe('facets', () => {
  it('simple string facet', async () => {
    const client = new Client({
      connection: {
        host: 'http://localhost:9200',
        apiKey: 'a2Rha1VJTUJMcGU4ajA3Tm9fZ0Y6MjAzX2pLbURTXy1hNm9SUGZGRlhJdw=='
      },
      search_settings: {
        highlight_attributes: ['title'],
        search_attributes: ['title'],
        result_attributes: ['title'],
        facet_attributes: ['rated', 'type']
      }
    })

    nock('http://localhost:9200')
      .post('/_msearch', (requestBody: any) => {
        expect(requestBody).toMatchSnapshot('ES Request')
        const x = JSON.parse(requestBody.split('\n')[1])

        expect(x.aggs).toMatchInlineSnapshot(`
          {
            "rated": {
              "terms": {
                "field": "rated",
                "size": 10,
              },
            },
            "type": {
              "terms": {
                "field": "type",
                "size": 10,
              },
            },
          }
        `)

        // has the nested filter applied
        expect(x.query).toMatchInlineSnapshot(`
          {
            "bool": {
              "filter": [
                {
                  "bool": {
                    "should": [
                      {
                        "term": {
                          "type": "movie",
                        },
                      },
                    ],
                  },
                },
              ],
              "must": {
                "bool": {
                  "should": [
                    {
                      "bool": {
                        "should": [
                          {
                            "multi_match": {
                              "fields": [
                                "title",
                              ],
                              "fuzziness": "AUTO:4,8",
                              "query": "shawshank",
                            },
                          },
                          {
                            "multi_match": {
                              "fields": [
                                "title",
                              ],
                              "query": "shawshank",
                              "type": "bool_prefix",
                            },
                          },
                        ],
                      },
                    },
                    {
                      "multi_match": {
                        "fields": [
                          "title",
                        ],
                        "query": "shawshank",
                        "type": "phrase",
                      },
                    },
                  ],
                },
              },
            },
          }
        `)
        return true
      })
      .reply(200, HitsResponseWithFacetFilter)

    const response = await client.handleInstantSearchRequests(
      DisjunctiveExampleRequest as AlgoliaMultipleQueriesQuery[]
    )

    expect(response).toMatchSnapshot()
  })
})
