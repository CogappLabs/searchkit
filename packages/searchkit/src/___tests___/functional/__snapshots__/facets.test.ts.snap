// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`facets simple string facet 3`] = `
{
  "results": [
    {
      "appliedRules": [],
      "exhaustive": {
        "facetsCount": true,
        "nbHits": true,
        "typo": true,
      },
      "exhaustiveFacetsCount": true,
      "exhaustiveNbHits": true,
      "exhaustiveTypo": true,
      "facets": {
        "actors": {
          "Bob Gunton": 1,
          "Morgan Freeman": 1,
          "Tim Robbins": 1,
          "William Sadler": 1,
        },
        "rated": {
          "R": 1,
        },
        "type": {
          "movie": 1,
        },
      },
      "facets_stats": {},
      "hits": [
        {
          "_highlightResult": {
            "title": {
              "fullyHighlighted": false,
              "matchLevel": "full",
              "matchedWords": [
                "Shawshank",
              ],
              "value": "The <em>Shawshank</em> Redemption",
            },
          },
          "_index": "imdb_movies",
          "_score": 7.637228,
          "actors": [
            "Tim Robbins",
            "Morgan Freeman",
            "Bob Gunton",
            "William Sadler",
          ],
          "objectID": "tt0111161",
          "title": "The Shawshank Redemption",
        },
      ],
      "hitsPerPage": 20,
      "index": "imdb_movies",
      "nbHits": 1,
      "nbPages": 1,
      "page": 0,
      "params": "facetFilters=type%3Amovie&facets=*&highlightPostTag=%3C%2Fem%3E&highlightPreTag=%3Cem%3E&maxValuesPerFacet=10&page=0&query=shawshank&tagFilters=",
      "processingTimeMS": 2,
      "query": "shawshank",
      "renderingContent": {
        "facetOrdering": {
          "facets": {
            "order": [
              "rated",
              "type",
            ],
          },
          "values": {
            "rated": {
              "sortRemainingBy": "count",
            },
            "type": {
              "sortRemainingBy": "count",
            },
          },
        },
      },
    },
    {
      "appliedRules": [],
      "exhaustive": {
        "facetsCount": true,
        "nbHits": true,
        "typo": true,
      },
      "exhaustiveFacetsCount": true,
      "exhaustiveNbHits": true,
      "exhaustiveTypo": true,
      "facets": {
        "type": {
          "movie": 1,
        },
      },
      "facets_stats": {},
      "hits": [
        {
          "_highlightResult": {
            "title": {
              "fullyHighlighted": false,
              "matchLevel": "full",
              "matchedWords": [
                "Shawshank",
              ],
              "value": "The <em>Shawshank</em> Redemption",
            },
          },
          "_index": "imdb_movies",
          "_score": 7.637228,
          "actors": [
            "Tim Robbins",
            "Morgan Freeman",
            "Bob Gunton",
            "William Sadler",
          ],
          "objectID": "tt0111161",
          "title": "The Shawshank Redemption",
        },
      ],
      "hitsPerPage": 1,
      "index": "imdb_movies",
      "nbHits": 1,
      "nbPages": 1,
      "page": 0,
      "params": "analytics=false&attributesToHighlight=&attributesToRetrieve=&attributesToSnippet=&clickAnalytics=false&facets=type&highlightPostTag=%3C%2Fem%3E&highlightPreTag=%3Cem%3E&hitsPerPage=1&maxValuesPerFacet=10&page=0&query=shawshank&tagFilters=",
      "processingTimeMS": 2,
      "query": "shawshank",
      "renderingContent": {
        "facetOrdering": {
          "facets": {
            "order": [
              "rated",
              "type",
            ],
          },
          "values": {
            "rated": {
              "sortRemainingBy": "count",
            },
            "type": {
              "sortRemainingBy": "count",
            },
          },
        },
      },
    },
  ],
}
`;

exports[`facets simple string facet: ES Request 1`] = `
"{"index":"imdb_movies"}
{"aggs":{"rated":{"terms":{"field":"rated","size":10}},"type":{"terms":{"field":"type","size":10}}},"query":{"bool":{"filter":[{"bool":{"should":[{"term":{"type":"movie"}}]}}],"must":{"bool":{"should":[{"bool":{"should":[{"multi_match":{"query":"shawshank","fields":["title"],"fuzziness":"AUTO:4,8"}},{"multi_match":{"query":"shawshank","fields":["title"],"type":"bool_prefix"}}]}},{"multi_match":{"query":"shawshank","type":"phrase","fields":["title"]}}]}}}},"size":20,"from":0,"_source":{"includes":["title"]},"highlight":{"pre_tags":["<em>"],"post_tags":["</em>"],"fields":{"title":{"number_of_fragments":0}}}}
{"index":"imdb_movies"}
{"aggs":{"type":{"terms":{"field":"type","size":10}}},"query":{"bool":{"filter":[],"must":{"bool":{"should":[{"bool":{"should":[{"multi_match":{"query":"shawshank","fields":["title"],"fuzziness":"AUTO:4,8"}},{"multi_match":{"query":"shawshank","fields":["title"],"type":"bool_prefix"}}]}},{"multi_match":{"query":"shawshank","type":"phrase","fields":["title"]}}]}}}},"size":1,"from":0,"_source":{"includes":["title"]},"highlight":{"pre_tags":["<em>"],"post_tags":["</em>"],"fields":{"title":{"number_of_fragments":0}}}}
"
`;
